### PR TITLE
Pause on recovery target in continuous recovery

### PR DESCRIPTION
**Pull request by roicos**
Previously when in archive recovery we were not able to pause
 on specific wal record. The recovery continued until it
 reached recovery_target, specified in postgresql.conf,
 which is not reloadable, so to continue recovery to the
 next recovery_target we had to stop the cluster, change recovery_target
 and start the cluster again.
 If we want to enable the cluster for read-only queries while
 continuing archive recovery, we can not restart the cluster
 to reload the recovery target. In this commit, we introduce the ability
 to pause in recovery without restarting the cluster.

 To achieve this the `gp_pause_on_restore_point_replay` GUC
 is changed from being just a bool on/off switch to a string
 that specifies the restore point name to pause on. Everything
 else stays the same except for when applying the restore point
 xlog record in xlog_redo() we now check if the restore point name is
 the one specified in the guc and pause the recovery.
 The `SELECT pg_wal_replay_resume();` call is required to continue
 recovery and to the next restore point specified or to promote.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/access/transam/xlog.c [hunk 1]
```diff
@@ -10837,10 +10837,15 @@ xlog_redo(XLogReaderState *record)
...
+		 * in gp_pause_on_restore_point_replay guc
+		 * WAL replay is paused at the restore point until it is
 		 * explicitly resumed.
 		 */
-		if (gp_pause_on_restore_point_replay)
+		xl_restore_point *recordRestorePointData;
+		recordRestorePointData = (xl_restore_point *) XLogRecGetData(record);
+
+		if (strcmp(recordRestorePointData->rp_name, gp_pause_on_restore_point_replay) == 0)
 			SetRecoveryPause(true);
```
Comment (soumyadeep2007) [inline src/backend/access/transam/xlog.c:10849]: I thought about this for a bit. There is a slight gotcha with the implementation of this GUC. You can call me pedantic/paranoid but just want to make us aware, in case we need to adjust our client application code.

TL;DR: After we replay a restore point we want to pause on, we read (not apply) one more WAL record before "pausing" (i.e. sleeping in `recoveryPausesHere()`)

## Existing workflow:

(0) The GUC is set and the config is reloaded.
(1) We replay a RP record and we check if its name matches our GUC.
(2) If it does, then we call `SetRecoveryPause()`, which sets the control variable: `XLogCtl->recoveryPause`
(3) Note that **(2) does NOT immediately pause recovery.** Instead, in the `main redo apply loop`, we will **read the next record** and then **eventually** pause here at:
```
/*
 * Pause WAL replay, if requested by a hot-standby session via
 * SetRecoveryPause().
```

## Observations:
* Even though we **won't replay the next WAL record**, we still have to "read" the next WAL record, i.e. call `ReadRecord`, which can hang arbitrarily if it is unavailable (worst case: RP was last WAL record in the current file and the next file is unobtainable for whatever reason).
* This probability of hanging is why upstream decided to make this [commit](https://github.com/postgres/postgres/commit/32fd2b57d7f64948e649fc205c43f007762ecaac). Also see this [discussion](https://www.postgresql.org/message-id/CAFiTN-sZz_aF0ny5NT8vt47N6%3DxTsvZ-%2B7vOQE3%3DYETO_fissA%40mail.gmail.com)
* So, if the next ReadRecord is stuck or taking time, the call to `recoveryPausesHere` will be delayed -> log line will be delayed and the sleep will be delayed.
    * Now this may not be a big deal, depending on the application code. Logically, it is true that replay is paused -> we may try to read the next record but we won't try to replay it. So it is "paused" in that sense, just not where we expect it to be.
* If the next `ReadRecord` hangs, and we call `pg_is_wal_replay_paused`, we will still get `true`.

## What if we made the pause synchronous?

```
if (strcmp(recordRestorePointData->rp_name, gp_pause_on_restore_point_replay) == 0)
{
			SetRecoveryPause(true);
            recoveryPausesHere();   <---- similar to case RECOVERY_TARGET_ACTION_PAUSE in switch (recoveryTargetAction) 
}
```

The above has a nasty side effect though (see action item below to confirm). We wouldn't be doing anything after this line:
```
/* Now apply the WAL record itself */
RmgrTable[record->xl_rmid].rm_redo(xlogreader);
```
till the end of the loop. This means we won't be advancing the `lastReplayedEndRecPtr`, which is what `pg_last_wal_replay_lsn` relies on.

I think this side effect is a deal breaker because **we need to inspect the replay location to know if we have replayed up to a certain restore point**.

## Action item:
Use the above patch with the added `recoveryPausesHere()` and see if using the restore point LSN directly from the return value of `pg_create_restore_point` instead of calling `pg_current_wal_lsn()`

## Middle ground:
Add `ereport(LOG, (errmsg("setting pause for restore point: '%s, created at time: %s', recordRestorePointData->rp_name, timestamptz_to_str(recordRestorePointData->rp_time))))`

Comment (roicos) [reply src/backend/access/transam/xlog.c:10849]: Tested this scenario as per discussion yesterday.
Without the patch the test still requires switch_wal, but the `XLogCtl->lastReplayedEndRecPtr` gets updated correctly when replaying. With the patch the `XLogCtl->lastReplayedEndRecPtr` shows the value < the actual LSN of restore point record and the test gets stuck waiting for `pg_last_wal_replay_lsn()` to return the value >= requested restore point lsn.

File: src/test/recovery/t/101_restore_point_and_startup_pause.pl [hunk 2]
```diff
@@ -1,48 +1,107 @@
...
+use Test::More tests => 10;
+
+my $node_primary;
+my $node_standby;
+
+sub test_pause_in_recovery
+{
+	my ($restore_point, $test_lsn, $num_rows) = @_;
+	# Wait until standby has replayed enough data
+	my $caughtup_query = "SELECT '$test_lsn'::pg_lsn <= pg_last_wal_replay_lsn()";
```
Comment (soumyadeep2007) [inline src/test/recovery/t/101_restore_point_and_startup_pause.pl:18]: I think we can have a tighter condition here. `<=` -> `=`:

`SELECT '$test_lsn'::pg_lsn = pg_last_wal_replay_lsn()`

Slightly easier to read:

`SELECT pg_last_wal_replay_lsn() = '$test_lsn'::pg_lsn`

File: src/backend/utils/misc/guc_gp.c [hunk 3]
```diff
@@ -4761,6 +4750,18 @@ struct config_string ConfigureNamesString_gp[] =
 	},
 #endif  /* ENABLE_IC_PROXY */
 
+	{
+		{"gp_pause_on_restore_point_replay", PGC_SIGHUP, DEVELOPER_OPTIONS,
```
Comment (soumyadeep2007) [inline src/backend/utils/misc/guc_gp.c:4754]: One little thing here. I guess we want only super-users to be able to change this GUC. So, we should go with `PGC_SUSET`. It is like `PGC_SIGHUP`, but only allows super-users to do the reload.

```
 * SUSET options can be set at postmaster startup, with the SIGHUP
 * mechanism, or from the startup packet or SQL if you're a superuser.
```

**Review by soumyadeep2007 — APPROVED** [review]
LGTM! Just have a couple of minor comments.

File: .abi-check/7.1.0/postgres.symbols.ignore [hunk 4]
```diff
@@ -5,3 +5,5 @@ has_update_triggers
 ConfigureNamesBool_gp
 aocs_beginscan
 AppendOnlyBlockDirectory_GetEntry
+ConfigureNamesString_gp
+gp_pause_on_restore_point_replay
```
Comment (jimmyyih) [inline .abi-check/7.1.0/postgres.symbols.ignore:9]: Need newline to get rid of the red symbol.

File: src/backend/access/transam/xlog.c [hunk 5]
```diff
@@ -10837,11 +10837,23 @@ xlog_redo(XLogReaderState *record)
 		/*
 		 * GPDB: Restore point records can act as a point of
 		 * synchronization to ensure cluster-wide consistency during WAL
-		 * replay. WAL replay is paused at each restore point until it is
+		 * replay. If the restore point is specified
+		 * in gp_pause_on_restore_point_replay guc
+		 * WAL replay is paused at the restore point until it is
```
Comment (jimmyyih) [inline src/backend/access/transam/xlog.c:10842]: Could be formatted better:
```
		 * replay. If the restore point is specified in the GUC
		 * gp_pause_on_restore_point_replay, WAL replay will be
		 * paused at the restore point until it is explicitly resumed.
```

File: src/backend/access/transam/xlog.c [hunk 6]
```diff
@@ -10837,11 +10837,23 @@ xlog_redo(XLogReaderState *record)
...
-		if (gp_pause_on_restore_point_replay)
+		xl_restore_point *recordRestorePointData;
+		recordRestorePointData = (xl_restore_point *) XLogRecGetData(record);
+
+		if (strcmp(recordRestorePointData->rp_name, gp_pause_on_restore_point_replay) == 0)
+		{
+			ereport(LOG,
+					(errmsg("setting pause for restore point: '%s', created at time: %s",
+							recordRestorePointData->rp_name,
+							timestamptz_to_str(recordRestorePointData->rp_time))));
```
Comment (jimmyyih) [inline src/backend/access/transam/xlog.c:10853]: Recommend changing the message:
```suggestion
					(errmsg("pausing on restore point '%s' created at on %s",
							recordRestorePointData->rp_name,
							timestamptz_to_str(recordRestorePointData->rp_time))));
```

But at the same time... why do we need to log out `rp_time`?  It doesn't seem too useful.

File: src/test/recovery/t/101_restore_point_and_startup_pause.pl [hunk 7]
```diff
@@ -1,48 +1,107 @@
-# test for pausing on startup and on a specified restore point
+# Test for pausing/resuming recovery a specific restore points, both at
+# initial startup and in a continuous fashion by shifting
+# gp_pause_on_restore_point_replay.
```
Comment (jimmyyih) [inline src/test/recovery/t/101_restore_point_and_startup_pause.pl:3]:
```suggestion
# Test for pausing and resuming recovery at specific restore points,
# both at initial startup and in a continuous fashion by advancing
# gp_pause_on_restore_point_replay.
```

File: src/test/recovery/t/101_restore_point_and_startup_pause.pl [hunk 8]
```diff
@@ -1,48 +1,107 @@
...
+test_pause_in_recovery('rp2', $lsn2, 3000);
 
-# Create a restore point on the primary
-my $restore_point_lsn =
-    $node_primary->safe_psql('postgres', "SELECT pg_create_restore_point('rp')");
+# go to rp3
+$node_standby->adjust_conf('postgresql.conf', 'gp_pause_on_restore_point_replay', "rp3");
+$node_standby->reload;
+$node_standby->safe_psql('postgres', "SELECT pg_wal_replay_resume();");
+test_pause_in_recovery('rp3', $lsn3, 4000);
```
Comment (jimmyyih) [inline src/test/recovery/t/101_restore_point_and_startup_pause.pl:89]: This last one doesn't seem necessary.  In fact, it might be better to remove it and validate below that the promotion happened exactly at rp2 and did not play any extra WAL records before exiting recovery and promoting.

File: src/backend/access/transam/xlog.c (hunk 6 above)
Comment (jimmyyih) [reply src/backend/access/transam/xlog.c:10853]: Talking with Deep, we'll keep the message as "setting pause on restore point '%s'" and take out the rp_time part.

File: src/test/recovery/t/101_restore_point_and_startup_pause.pl (hunk 8 above)
Comment (jimmyyih) [reply src/test/recovery/t/101_restore_point_and_startup_pause.pl:89]: Discussing with Deep, we'll keep rp3 and also add two more restore point locations to handle a skip-over scenario and the promotion scenario that I mentioned in my last comment.

Comment (jimmyyih) [reply src/test/recovery/t/101_restore_point_and_startup_pause.pl:89]: After adding the extra two test scenarios, I discovered that this PR's patch does not handle the promotion correctly.  I will close this PR and open up a new PR with a different better approach.

**Comment by jimmyyih** [comment]
Closed in favor of https://github.com/greenplum-db/gpdb/pull/17339.
